### PR TITLE
emit space before the dot for property/element access expressions if …

### LINKED ISSUE #4155: TypeScript 1.5 const enum changes break enum.toString()
I followed the [recommendation](https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#enum-reference-in-lining-changes) by changing my `.d.ts` files to use `const` for enum definitions (these are generated view Web Essentials plugin from server-side classes.

Now I am able to compile but I get run time errors when I try to use toString on a number literal. Previously this was not a problem because the compiler put in a comment (which I liked btw) and a space. Here is some example code to reproduce the issue.

```ts
// FileId.d.ts
declare module server {
	const enum FieldId {
		ABSTRACT = 7000,
		RANK = 40002
	}
}

// Usage
var f = getById(server.FieldId.RANK.toString());

// Previous output
var f = getById(4002 /* RANK */.toString());

// TypeScript 1.5 output
var f = getById(4002.toString());
```

This is disappointing because TypeScript is supposed to catch errors like this at compile time. Should this be a compiler error in 1.5? What is the suggested solution?

### PR DESCRIPTION
**Pull request by vladima**
…expression part is integer const enum. Fixes #4155

### CODE REVIEW

File: tests/cases/compiler/constEnumToStringWithComments.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// @comments: yes
```
Comment (RyanCavanaugh) [inline tests/cases/compiler/constEnumToStringWithComments.ts]: Do we actually support `yes` here?

**Comment by RyanCavanaugh** [comment]
:+1:

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2054,10 +2059,19 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                if (!indentedBeforeDot) {
+                    if (node.expression.kind === SyntaxKind.NumericLiteral) {
+                        // check if numeric literal was originally written with a dot
+                        let text = getSourceTextOfNodeFromSourceFile(currentSourceFile, node.expression);
+                        shouldEmitSpace = text.indexOf(tokenToString(SyntaxKind.DotToken)) < 0;
+                    }
+                    else {
+                        // check if constant enum value is integer
+                        let constantValue = tryGetConstEnumValue(node.expression);
+                        shouldEmitSpace = constantValue !== undefined && isFinite(constantValue) && Math.floor(constantValue) === constantValue;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: You can remove the `constantValue !== undefined` check

File: tests/cases/compiler/constEnumToStringWithComments.ts (hunk 1 above)
Comment (vladima) [inline tests/cases/compiler/constEnumToStringWithComments.ts]: oops, anything but `false` means `true` here. Will fix it

File: src/compiler/emitter.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/emitter.ts]: ok

File: tests/cases/compiler/constEnumToStringWithComments.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline tests/cases/compiler/constEnumToStringWithComments.ts]: Can we make the harness error on seeing a value other than `true` or `false`?

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**